consensus/src/marshal/actor.rs

```diff
@@ -979,97 +978,53 @@ impl<
-                    let finalization = self.cache.get_finalization_for(commitment).await;
-                    self.store_finalization(
-                        block.height(),
-                        commitment,
-                        block.clone(),
-                        finalization,
-                        application,
-                    )
-                    .await;
-                    debug!(height = block.height(), "repaired block");
-                    cursor = block;
-                } else {
-                    // Request the next missing block digest
-                    resolver.fetch(Request::<B>::Block(commitment)).await;
-                    break;
-                }
-            }
 
-            // If we haven't fully repaired the gap, then also request any possible
-            // finalizations for the blocks in the remaining gap. This may help
-            // shrink the size of the gap if finalizations for the requests heights
-            // exist. If not, we rely on the recursive digest fetch above.
-            let end = std::cmp::max(cursor.height(), start);
-            debug!(start, end, "requesting any finalized blocks");
-            let requests = (start..end)
-                .map(|height| Request::<B>::Finalized { height })
-                .collect();
-            resolver.fetch_all(requests).await;
-        }
-    }
+        // Attempt to repair the gap backwards from the end of the gap, using
+        // blocks from our local storage.
+        let Some(mut cursor) = self.get_finalized_block(gap_end).await else {
+            panic!("gapped block missing that should exist: {gap_end}");
+        };
 
-    /// Identifies one or more of the earliest gaps in the finalized blocks archive. The gaps
-    /// returned are half-open ranges, where `start` is inclusive and `end` is exclusive. The total
-    /// number of missing heights covered by the returned gaps is bounded by `self.max_repair`.
-    ///
-    /// The last gap may in the returned list be partial, i.e., it may cover fewer heights
-    /// than the full gap size, if the remaining number of heights to repair exceeds the
-    /// `max_repair` limit.
-    fn identify_gaps(&self) -> Vec<Range<u64>> {
-        // Ignore ranges below the last processed height
-        let start = self.last_processed_height.saturating_add(1);
-        // The remaining number of items to repair
-        let mut rem = self.max_repair.get();
-        // All gaps must be before a range, so we iterate over the ranges and collect the gaps.
-        // Remember that (s, e) are the start and end of the next full range, but we want to collect the gaps.
-        let mut gaps = Vec::new();
-        for (s, _) in self.finalized_blocks.ranges() {
-            // Break if finished
-            if rem == 0 {
+        // Iterate backwards, repairing blocks as we go.
+        while cursor.height() > next_needed {
```

Comment:
If we end up repairing the entire gap from locally-sourced blocks, I think we should loop and call `self.finalized_blocks.next_gap()` again to find the NEXT possible gap that we can iterate over